src/libsystemd-network/ndisc-router.c

```diff
@@ -715,3 +715,29 @@ int sd_ndisc_router_dnssl_get_lifetime(sd_ndisc_router *rt, uint32_t *ret_sec) {
         *ret_sec = be32toh(*(uint32_t*) (ri + 4));
         return 0;
 }
+
+int sd_ndisc_router_captive_portal_get_uri(sd_ndisc_router *rt, const char **uri, size_t *size) {
+        int r;
+        const char *nd_opt_captive_portal;
+        size_t length;
+
+        assert_return(rt, -EINVAL);
+        assert_return(uri, -EINVAL);
+
+        r = sd_ndisc_router_option_is_type(rt, SD_NDISC_OPTION_CAPTIVE_PORTAL);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return -EMEDIUMTYPE;
+
+        r = sd_ndisc_router_option_get_raw(rt, (void *)&nd_opt_captive_portal, &length);
+        if (r < 0)
+                return r;
+        if (length < 2)
+                return -EBADMSG;
+
+        *uri = nd_opt_captive_portal + 2;
+        *size = strnlen(*uri, length);
```

Comment:
length -> length - 2

Maybe, better to return `NULL` when `length == 2` ??

Reply:
> length -> length - 2

Ah, yep.

> Maybe, better to return NULL when length == 2 ??

Yeah, probably. The dhcp4/6 cases try not to record empty strings at least, which I chose based on the existing behavior of `dhcp_option_parse_string()`. I will make all three options uniformly return NULL in case of empty strings.